polygeist/lib/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.cpp

```diff
@@ -0,0 +1,196 @@
+      if (currentModifiers != newModifiers) {
+        currentModifiers.insert(newModifiers.begin(), newModifiers.end());
+        result |= ChangeResult::Change;
+      }
+    }
+  };
+
+  join(valueToModifiers, otherReachingDef.valueToModifiers);
+  join(valueToPotentialModifiers, otherReachingDef.valueToPotentialModifiers);
+  return result;
+}
+
+ChangeResult ReachingDefinition::reset() {
+  if (valueToModifiers.empty() && valueToPotentialModifiers.empty())
+    return ChangeResult::NoChange;
+
+  valueToModifiers.clear();
+  valueToPotentialModifiers.clear();
+  return ChangeResult::Change;
+}
+
+ChangeResult ReachingDefinition::setModifier(Value val, Operation *op) {
+  ReachingDefinition::ModifiersTy &mods = valueToModifiers[val];
+  if (mods.size() == 1 && mods.front() == op)
+    return ChangeResult::NoChange;
+
+  // Set the new modifier and clear out all previous definitions.
+  mods.clear();
+  mods.insert(op);
+  valueToPotentialModifiers[val].clear();
+  return ChangeResult::Change;
+}
+
+ChangeResult ReachingDefinition::addPotentialModifier(Value val,
+                                                      Operation *op) {
+  valueToPotentialModifiers[val].insert(op);
+  return ChangeResult::Change;
+}
+
+std::optional<ArrayRef<Operation *>>
+ReachingDefinition::getModifiers(Value val) const {
+  auto it = valueToModifiers.find(val);
+  return (it != valueToModifiers.end()) ? it->second.getArrayRef()
+                                        : std::nullopt;
+}
+
+std::optional<ArrayRef<Operation *>>
+ReachingDefinition::getPotentialModifiers(Value val) const {
+  auto it = valueToPotentialModifiers.find(val);
+  return (it != valueToPotentialModifiers.end()) ? it->second.getArrayRef()
+                                                 : std::nullopt;
+}
+
+//===----------------------------------------------------------------------===//
+// ReachingDefinitionAnalysis
+//===----------------------------------------------------------------------===//
+
+void ReachingDefinitionAnalysis::setToEntryState(ReachingDefinition *lattice) {
+  /// Set the initial state (nothing is known about reaching definitions).
+  propagateIfChanged(lattice, lattice->reset());
+}
+
+void ReachingDefinitionAnalysis::visitOperation(
+    Operation *op, const ReachingDefinition &before,
+    ReachingDefinition *after) {
+  LLVM_DEBUG(llvm::dbgs() << "ReachingDefinitionAnalysis - Visit: " << *op
+                          << "\n");
+
+  // Initialize the alias queries for the current function.
+  if (isa<FunctionOpInterface>(op)) {
+    auto funcOp = cast<FunctionOpInterface>(*op);
+    aliasQueriesMap[funcOp] =
+        std::make_unique<AliasQueries>(funcOp, aliasAnalysis);
+  }
+
+  // If an operation has unknown memory effects assume we can't deduce
+  // anything about reaching definitions.
+  auto memoryEffectOp = dyn_cast<MemoryEffectOpInterface>(op);
+  if (!memoryEffectOp) {
+    LLVM_DEBUG(llvm::dbgs() << "Operation has unknown side effects\n");
```

Comment:
Can we have one test case for this? I wonder what would the result look like. empty maps?